runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java

```diff
@@ -0,0 +1,232 @@
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Object & serialization methods for in-stream checkpoint
+ * summarization of SMR object state.
+ */
+@ToString(callSuper = true)
+@NoArgsConstructor
+public class CheckpointEntry extends LogEntry {
+
+    @RequiredArgsConstructor
+    public enum CheckpointEntryType {
+        START(1),           // Mandatory: 1st record in checkpoint
+        CONTINUATION(2),    // Optional: 2nd through (n-1)th record
+        END(3);             // Mandatory: final record checkpoint
+
+        public final int type;
+
+        public byte asByte() {
+            return (byte) type;
+        }
+    };
+
+    /**
+     * Constants for use as keys in our 'dict' map.
+     */
+    public static String START_TIME = "Start time";
+    public static String END_TIME = "End time";
+    public static String START_LOG_ADDRESS = "Start log address";
+    public static String ENTRY_COUNT = "Entry count";
+    public static String BYTE_COUNT = "Byte count";
+
+    /** Type of entry
+     */
+    @Getter
+    CheckpointEntryType cpType;
+
+    /**
+     * Unique identifier for this checkpoint.  All entries
+     * for the same checkpoint state must use the same ID.
+     */
+    @Getter
+    UUID checkpointID;
+
+    /** Author/cause/trigger of this checkpoint
+     */
+    @Getter
+    String checkpointAuthorID;
+
+    /** Map of checkpoint metadata, see key constants above
+     */
+    @Getter
+    Map<String, String> dict;
+
+    /** Optional: array of SMREntry objects that contains
+     *  SMR object state of the stream that we're checkpointing.
+     *  May be present in any CheckpointEntryType, but typically
+     *  used by CONTINUATION entries.
+     */
+    @Getter @Setter
+    SMREntry[] smrEntries = null;
+
+    /** Byte count of smrEntries in serialized form, zero
+     *  if smrEntries.size() is zero or if value is unknown.
+     */
+    @Getter
+    int smrEntriesBytes = 0;
+
+    public CheckpointEntry(CheckpointEntryType type, String authorID, UUID checkpointID,
+                           Map<String,String> dict, SMREntry[] smrEntries) {
+        super(LogEntryType.CHECKPOINT);
+        this.cpType = type;
+        this.checkpointID = checkpointID;
+        this.checkpointAuthorID = authorID;
+        this.dict = dict;
+        this.smrEntries = smrEntries;
+    }
+
+    static final Map<Byte, CheckpointEntryType> typeMap =
+            Arrays.stream(CheckpointEntryType.values())
+                    .collect(Collectors.toMap(CheckpointEntryType::asByte, Function.identity()));
+
+    /**
+     * This function provides the remaining buffer. Child entries
+     * should initialize their contents based on the buffer.
+     *
+     * @param b The remaining buffer.
+     * @param rt The CorfuRuntime used by the SMR object.
+     * @return A CheckpointEntry.
+     */
+    @Override
+    void deserializeBuffer(ByteBuf b, CorfuRuntime rt) {
```

Comment:
Hrm, I'd prefer not to.  The `SMREntry` items inside a `CheckpointEntry` have tangle problems with entry & runtime references that need fixing up in various places, similar to https://github.com/CorfuDB/CorfuDB/blob/6a10e5c/runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java#L45-L46 , and I'd like to avoid needing yet another case of doing that.  I hope that moving key checkpoint metadata to the top-level metadata map will be sufficient; if not, then I'll return here.